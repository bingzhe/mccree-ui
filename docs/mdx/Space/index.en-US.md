---
title: Space
mobile: true
nav:
    path: /components
---

Space en-us
