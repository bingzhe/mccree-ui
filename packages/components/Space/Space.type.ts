import { CSSProperties, ReactNode } from "react";

import { SizeType } from "../../../components/config-provider/SizeContext";

type DirectionType = "horizontal" | "vertical";

// export interface SpaceProps extends React.HTMLAttributes<HTMLDivElement> {
//     size?: SizeType | number;
//     direction?: DirectionType;
// }

export const defaultProps = {
    size: "small" as SizeType | number,
    direction: "horizontal" as DirectionType
};

export interface SpaceProps {
    style?: CSSProperties;
    className?: string;
    /**
     * @zh 对齐方式
     */
    align?: "start" | "end" | "center" | "baseline";
    /**
     * @zh 间距方向
     * @defaultValue horizontal
     */
    direction?: "horizontal" | "vertical";
    /**
     * @zh 尺寸
     * @defaultValue default
     */
    size?: "small" | "default" | "large";
    /**
     * @zh 换行。
     */
    wrap?: boolean;
    split?: ReactNode;
    children?: ReactNode;
}
