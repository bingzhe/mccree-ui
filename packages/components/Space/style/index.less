@import "../../styles/vars.less";

@space-prefix-cls: ~"@{mccree-prefix}-space";

.@{space-prefix-cls} {
    display: inline-flex;

    &-horizontal {
        align-items: center;
    }
    &-vertical {
        flex-direction: column;
    }
}
