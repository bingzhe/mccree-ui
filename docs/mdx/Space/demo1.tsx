import React from "react";
import { Space } from "mccree-ui/components";
import "@mccree-ui/components/Space/style/index";

export default () => {
    return (
        <Space>
            <div style={{ width: "50px", height: "50px", border: "1px solid #ccc" }}>1</div>
            <div style={{ width: "50px", height: "50px", border: "1px solid #ccc" }}>2</div>
            <div style={{ width: "50px", height: "50px", border: "1px solid #ccc" }}>3</div>
            <div style={{ width: "50px", height: "50px", border: "1px solid #ccc" }}>4</div>
        </Space>
    );
};
