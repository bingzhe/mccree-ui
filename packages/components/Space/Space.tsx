import React, { useContext } from "react";
import cs from "@mccree-ui/util/classnames";
import { ConfigContext } from "../ConfigProvider";
import { SpaceProps } from "./Space.type";
import { useMergeProps } from "@mccree-ui/hooks";

const spaceSize = {
    small: 8,
    medium: 16,
    large: 24
};

const defaultProps: SpaceProps = {
    size: "small",
    direction: "horizontal"
};

const Space: React.ForwardRefRenderFunction<unknown, SpaceProps> = (baseProps, ref) => {
    const { getPrefixCls, componentConfig } = useContext(ConfigContext);
    const props = useMergeProps(baseProps, defaultProps, componentConfig?.Space);
    const { className, size, direction, children, ...rest } = props;

    const prefixCls = getPrefixCls("space");

    const ChildrenList = React.Children.toArray(children);
    const len = ChildrenList.length;

    // const { getPrefixCls } = React.useContext(ConfigContext);

    const classes = cs(prefixCls, className, `${prefixCls}-${direction}`);
    const itemClassName = `${prefixCls}-item`;

    return (
        <div className={classes} {...rest}>
            {ChildrenList.map((child, i) => (
                <div
                    className={itemClassName}
                    key={`${itemClassName}-${i}`}
                    style={
                        i === len - 1
                            ? {}
                            : {
                                  [direction === "vertical" ? "marginBottom" : "marginRight"]:
                                      typeof size === "string" ? spaceSize[size] : size
                              }
                    }
                >
                    {child}
                </div>
            ))}
        </div>
    );
};

Space.displayName = "MR_Space";
Space.defaultProps = defaultProps;

export default Space;
