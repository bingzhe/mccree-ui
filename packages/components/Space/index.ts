import Space from "./Space";
import { SpaceProps } from "./Space.type";

export default Space;
export { SpaceProps };
