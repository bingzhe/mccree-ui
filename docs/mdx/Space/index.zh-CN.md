---
title: Space 间距
mobile: true
nav:
    path: /components
---

## Space

```tsx
import React from "react";
import { Space } from "mccree-ui/components";

// import "@mccree-ui/components/Space/style/index";

const a = () => {
    return <Space>default</Space>;
};

export default a;
```
