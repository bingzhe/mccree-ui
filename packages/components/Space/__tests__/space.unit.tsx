import * as React from "react";
import { render, mount } from "enzyme";
import toJson from "enzyme-to-json";

import Space from "../index";

describe("Space", () => {
    it("should render width empty children", () => {
        const wrapper = mount(<Space />);

        expect(wrapper.instance()).toBe(null);
    });

    it("should render width customize size", () => {
        const wrapper = mount(
            <Space size={10}>
                <span>1</span>
                <span>2</span>
            </Space>
        );

        expect((wrapper.find("div.mccree-space-item").at(0) as any).prop("style").marginRight).toBe(
            10
        );
        expect(
            (wrapper.find("div.mccree-space-item").at(1) as any).prop("style").marginRight
        ).toBeUndefined();
    });

    it("should render correct with children", () => {
        const wrapper = render(
            <Space>
                text1<span>text1</span>
                text3
            </Space>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
