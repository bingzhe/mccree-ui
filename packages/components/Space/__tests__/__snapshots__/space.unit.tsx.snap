// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Space should render correct with children 1`] = `
<div
  class="mccree-space mccree-space-horizontal"
>
  <div
    class="mccree-space-item"
    style="margin-right:8px"
  >
    text1
  </div>
  <div
    class="mccree-space-item"
    style="margin-right:8px"
  >
    <span>
      text1
    </span>
  </div>
  <div
    class="mccree-space-item"
  >
    text3
  </div>
</div>
`;
